### PR TITLE
feat(Harmonizer): compress pitch array to required length (#45)

### PR DESCRIPTION
**Pull request by MalteRichert**
resolves issue #45 
@ArtInLines I've put the soundLength-code into StateManager.testSound() with instructions, didn't actually add the call to Harmonizer as that would require mock data.
@nichtLehdev @Rex2002 please check whether this solves our problem

### CODE REVIEW

**Review by ArtInLines — CHANGES_REQUESTED** [review]
@MalteRichert I addeed test-data and found a bug. I pushed my changes with the test-data to this branch, so you can see the bug for yourself too.
The bug happens in `quantizePitch`. In the 4th iteration of the for-loop, `i * bufferLength + bufferLength` happens to be equal to `notes.length + bufferLength`. This means, that `excess == bufferLength` which in turn means, that you effectively try to copy from the range `i * bufferLength` to `i *bufferLength` leaving the actual buffer empty. Thus the runtime-exception you call below gets fired.
I didn't spend any time thinking about the underlying problem yet, but I feel like your idea behind how to calculate the range to copy from the Pitch is kinda fundamentally flawed.

**Comment by ArtInLines** [comment]
I am not entirely sure I understand what you are planning to do with this function too. Is your goal just to split the pitch-array into as many chunks as you have numberBeats and then transform each of those chunks into its average?
Because then I would just do the following:
```java
double[] notes = new double[numberBeats];
int bufferLength = pitch.length / numberBeats;
for (int i = 0, bufferStart = 0; i < notes.length; i++, bufferStart += bufferLength) {
    notes[i] = 0;
    for (int j = bufferStart; j < bufferStart + bufferLength; j++) {
        notes[i] += pitch[j] / bufferLength;
    }
}
return notes;
```

**Comment by Rex2002** [comment]
I don't quite understand the idea behind theaveraging-approach by @MalteRichert, but I think using @ArtInLines code for the averaging, should yield the results we were hoping to get, assuming that our data is always more plentiful than the number of notes required. 
Do we intend to rely on that or are we going to implement another function to handle the case when our data is shorter than the number of notes required?

**Comment by nichtLehdev** [comment]
I agree with @Rex2002 on the results this approach should yield. I think that this could be the restricting part of the program, additionally to the Chart Analysis, which sets the minimum data required. 
I think we do already check for that somewhere. (didn't find it directly while skipping through code tho)

As long as we check that in Chart Analysis or State Manager, we should be fine in the Harmonizer. 
Else we should probably write a check here.

**Comment by MalteRichert** [comment]
Implemented @ArtInLines 's code so quantisation should work now. I already noted in javadoc that quantizePitch assumes that the pitch array is much longer than the resulting notes array. I agree with @nichtLehdev that that should be required prerequisite for data density.

I noticed while testing (really great that we can test now by the way!!) that pitch normalisation doesn't seem to work though, as the diverse set of TestDoubles led to an array of 90 times the same note. Will open a separate issue for that.

File: src/main/java/app/mapping/InstrumentDataRaw.java [hunk 1]
```diff
@@ -6,30 +6,50 @@
...
-    boolean[] onOffFilter;
+    public double[] frequency;
+    public boolean highPass;
+    public boolean[] onOffFilter;
 
     // Panning
-    double[] pan;
+    public double[] pan;
+
+    public InstrumentDataRaw(double[] relVolume, boolean[] absVolume, double[] pitch, InstrumentEnum instrument,
```
Comment (MalteRichert) [inline src/main/java/app/mapping/InstrumentDataRaw.java:36]: @ArtInLines why did you add a constructor here? Shouldn't Lombok provide exactly this constructor via @Data?

File: src/main/java/audio/harmonizer/Harmonizer.java [hunk 2]
```diff
@@ -63,10 +68,14 @@ private int[] normalizePitch(double[] pitch) throws AppError {
 
     private double[] getRandomScale() {
         double[] scale = new double[] {
-                2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f,
-                2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f,
-                2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f,
-                2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f, 1/76f, 2/76f
+                2 / 76f, 1 / 76f, 2 / 76f, 1 / 76f, 2 / 76f, 2 / 76f, 1 / 76f, 2 / 76f, 1 / 76f, 2 / 76f, 1 / 76f,
```
Comment (MalteRichert) [inline src/main/java/audio/harmonizer/Harmonizer.java:71]: @ArtInLines Is this autoLint or do you actually prefer this? I think it's way more readable as compact fractions

Comment (nichtLehdev) [reply src/main/java/audio/harmonizer/Harmonizer.java:71]: Looks like autoLint.. Also prefer the compact option. Would recall that change.

**Review by nichtLehdev — APPROVED** [review]
I'm fine with approving this in its current state. Maybe we should reformat our fraction array. But that's a minor change which only helps to read the code a bit better.

**Review by ArtInLines — APPROVED** [review]
